### ISSUE TITLE
[RISCV] Miscompile at -O3 with zbb

### ISSUE THREAD

**Issue by dtcxzyw** [llvm/llvm-project#85190]
Reduced test case:
```
#include <stdint.h>
#include <stdio.h>
int8_t d;
int32_t b;
uint64_t c;
int64_t safe_add_func_int64_t_s_s(int64_t si1, int64_t si2)
{
  return ((si1<0) && (si2<0) && (si1 < (INT64_MIN-si2))) ?
    (si1) :
    (si1 + si2);
}
int main() {
  do b = safe_add_func_int64_t_s_s(c, 7 | c);
  while (d);
  printf("%d\n", b);
  return 0;
}
```

```
> gcc -O0 test.c -fsanitize=undefined,address && ./a.out
7
> bin/clang -O3 -march=rv64gc_zbb test.c --target=riscv64-linux-gnu
> qemu-riscv64 -L /usr/riscv64-linux-gnu -cpu rv64,zbb=true ./a.out
0
```

LLVM version: eb21ee49cff081911d99d29ba887c1715fc2b8fc
qemu version: v8.2.1

cc @topperc @asb

**Comment by dtcxzyw** [comment]
https://godbolt.org/z/7cqMeaox9

**Comment by dtcxzyw** [comment]
Reduced IR: https://godbolt.org/z/f8Th3dhPs
```
define i1 @test1(i64 %a) {
  %or = or i64 %a, 7
  %or.cond.not.i = icmp slt i64 %a, 0
  %sub.i = sub nsw i64 -9223372036854775808, %or
  %cmp3.i = icmp sgt i64 %sub.i, %a
  %or.cond.i = select i1 %or.cond.not.i, i1 %cmp3.i, i1 false
  ret i1 %or.cond.i
}
```

**Comment by dtcxzyw** [comment]
Alive2: https://alive2.llvm.org/ce/z/QdRgGh

**Comment by dtcxzyw** [comment]
Should be fixed by #84924.
